Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -292,6 +304,84 @@ inline bool _ShouldDrawToCommandList() {
     HRESULT _CreateShadowEffect(ID2D1Image* inputImage, ID2D1Effect** outShadowEffect);
 
 public:
+    inline D2D1_ANTIALIAS_MODE GetAntialiasMode() {
+        if (CurrentGState().shouldAntialias == _kCGToggleDefault && allowsAntialiasing == _kCGToggleDefault) {
+            return D2D1_ANTIALIAS_MODE_PER_PRIMITIVE;
+        }
+
+        return CurrentGState().shouldAntialias == _kCGToggleOn && allowsAntialiasing == _kCGToggleOn ? D2D1_ANTIALIAS_MODE_PER_PRIMITIVE :
+                                                                                                       D2D1_ANTIALIAS_MODE_ALIASED;
+    }
+
+    inline D2D1_TEXT_ANTIALIAS_MODE GetTextAntialiasMode() {
+        if (CurrentGState().shouldSmoothFonts == _kCGToggleDefault && allowsFontSmoothing == _kCGToggleDefault) {
+            return D2D1_TEXT_ANTIALIAS_MODE_DEFAULT;
+        }
+
+        if (CurrentGState().shouldAntialias == _kCGToggleOn && allowsAntialiasing == _kCGToggleOn &&
+            CurrentGState().shouldSmoothFonts == _kCGToggleOn && allowsFontSmoothing == _kCGToggleOn) {
+            return D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE;
+        } else {
+            return D2D1_TEXT_ANTIALIAS_MODE_GRAYSCALE;
+        }
+
+        return D2D1_TEXT_ANTIALIAS_MODE_ALIASED;
+    }
+
+    inline DWRITE_RENDERING_MODE GetRenderingMode() {
+        if (CurrentGState().shouldAntialias == _kCGToggleDefault && allowsAntialiasing == _kCGToggleDefault) {
+            return DWRITE_RENDERING_MODE_DEFAULT;
+        }
+
+        if (CurrentGState().shouldAntialias == _kCGToggleOn && allowsAntialiasing == _kCGToggleOn &&
+            CurrentGState().shouldSubpixelPosition == _kCGToggleOn && allowsFontSubpixelPositioning == _kCGToggleOn) {
+            if (CurrentGState().shouldSubpixelQuantizeFonts == _kCGToggleOn && allowsFontSubpixelQuantization == _kCGToggleOn) {
+                return DWRITE_RENDERING_MODE_NATURAL_SYMMETRIC;
+            }
+            return DWRITE_RENDERING_MODE_NATURAL;
+        }
+        return DWRITE_RENDERING_MODE_DEFAULT;
+    }
+
+    inline IDWriteRenderingParams* GetTextRenderingParams(IDWriteRenderingParams* originalParams) {
+        IDWriteRenderingParams* workingParams = originalParams;
+
+        ComPtr<IDWriteFactory> dwriteFactory;
+        if (FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory))) {
+            return originalParams;
+        }
+        if (!originalParams) {
+            dwriteFactory->CreateRenderingParams(&workingParams);
+        }
+
+        IDWriteRenderingParams* textRenderingParams;
+
+        dwriteFactory->CreateCustomRenderingParams(workingParams->GetGamma(),
+                                                   workingParams->GetEnhancedContrast(),
+                                                   workingParams->GetClearTypeLevel(),
+                                                   workingParams->GetPixelGeometry(),
```

Comment:
There is no direct mapping between how Mac does subpixel positioning and Dwrite does it.  From the "intention" standpoint, subpixel positioning should map to clear type.